aequilibrae/paths/linear_approximation.py

```diff
@@ -0,0 +1,370 @@
+            aon_minus_current_sol = np.sum(c._aon_results.link_loads[:, :] - c.results.link_loads[:, :], axis=1)
+            aon_minus_prev_dir = np.sum(c._aon_results.link_loads[:, :] - stp_dir.link_loads[:, :], axis=1)
+            numerator += prev_dir_minus_current_sol * aon_minus_current_sol
+            denominator += prev_dir_minus_current_sol * aon_minus_prev_dir
+
+        numerator = np.sum(numerator * self.vdf_der)
+        denominator = np.sum(denominator * self.vdf_der)
+
+        alpha = numerator / denominator
+        if alpha < 0.0:
+            self.stepdirection = 0.0
+        elif alpha > self.conjugate_direction_max:
+            self.stepdirection = self.conjugate_direction_max
+        else:
+            self.conjugate_stepsize = alpha
+
+    def calculate_biconjugate_direction(self):
+        self.vdf.apply_derivative(
+            self.vdf_der, self.fw_total_flow, self.capacity, self.free_flow_tt, *self.vdf_parameters
+        )
+
+        # TODO: This should be a sum over all supernetwork links, it's not tested for multi-class yet
+        # if we can assume that all links appear in the subnetworks, then this is correct, otherwise
+        # this needs more work
+        mu_numerator = 0.0
+        mu_denominator = 0.0
+        nu_nom = 0.0
+        nu_denom = 0.0
+        for c in self.traffic_classes:
+            x_ = np.sum((self.step_direction[c.mode].link_loads[:, :] * self.stepsize
+                         + self.previous_step_direction[c.mode].link_loads[:, :] * (1.0 - self.stepsize)
+                         - c.results.link_loads[:, :]), axis=1)
+
+            y_ = np.sum(c._aon_results.link_loads[:, :] - c.results.link_loads[:, :], axis=1)
+            z_ = np.sum(self.step_direction[c.mode].link_loads[:, :] - c.results.link_loads[:, :], axis=1)
+            mu_numerator += x_ * y_
+            mu_denominator += x_ * np.sum(
+                self.previous_step_direction[c.mode].link_loads - self.step_direction[c.mode].link_loads[:, :], axis=1)
+            nu_nom += z_ * y_
+            nu_denom += z_ * z_
+
+        mu_numerator = np.sum(mu_numerator * self.vdf_der)
+        mu_denominator = np.sum(mu_denominator * self.vdf_der)
+        if mu_denominator == 0.0:
+            mu = 0.0
+        else:
+            mu = -mu_numerator / mu_denominator
+            # logger.info("mu before max = {}".format(mu))
+            mu = max(0.0, mu)
+
+        nu_nom = np.sum(nu_nom * self.vdf_der)
+        nu_denom = np.sum(nu_denom * self.vdf_der)
+        if nu_denom == 0.0:
+            nu = 0.0
+        else:
+            nu = -(nu_nom / nu_denom) + mu * self.stepsize / (1.0 - self.stepsize)
+            # logger.info("nu before max = {}".format(nu))
+            nu = max(0.0, nu)
+
+        self.betas[0] = 1.0 / (1.0 + nu + mu)
+        self.betas[1] = nu * self.betas[0]
+        self.betas[2] = mu * self.betas[0]
+
+    def calculate_step_direction(self):
```

Comment:
Just saw the "not intended for direct user" change, should we follow pep8 and use leading underscores for method names that are not supposed to be public?